tikv/raftstore/fsm_apply.go

```diff
@@ -833,26 +840,215 @@ func (d *applyDelegate) execRaftCmd(aCtx *applyContext, req *raft_cmdpb.RaftCmdR
+	case raft_cmdpb.AdminCmdType_BatchSplit:
+		adminResp, result, err = d.execBatchSplit(aCtx, adminReq)
+	case raft_cmdpb.AdminCmdType_CompactLog:
+		adminResp, result, err = d.execCompactLog(aCtx, adminReq)
+	case raft_cmdpb.AdminCmdType_TransferLeader:
+		err = errors.New("transfer leader won't execute")
+	case raft_cmdpb.AdminCmdType_ComputeHash:
+		adminResp, result, err = d.execComputeHash(aCtx, adminReq)
+	case raft_cmdpb.AdminCmdType_VerifyHash:
+		adminResp, result, err = d.execVerifyHash(aCtx, adminReq)
+	case raft_cmdpb.AdminCmdType_PrepareMerge:
+		adminResp, result, err = d.execPrepareMerge(aCtx, adminReq)
+	case raft_cmdpb.AdminCmdType_CommitMerge:
+		adminResp, result, err = d.execCommitMerge(aCtx, adminReq)
+	case raft_cmdpb.AdminCmdType_RollbackMerge:
+		adminResp, result, err = d.execRollbackMerge(aCtx, adminReq)
+	case raft_cmdpb.AdminCmdType_InvalidAdmin:
+		err = errors.New("unsupported command type")
+	}
+	if err != nil {
+		return
+	}
+	adminResp.CmdType = cmdType
+	resp = newCmdRespForReq(req)
+	resp.AdminResponse = adminResp
+	return
 }
 
 func (d *applyDelegate) execWriteCmd(aCtx *applyContext, req *raft_cmdpb.RaftCmdRequest) (
 	resp *raft_cmdpb.RaftCmdResponse, result applyResult, err error) {
-	return // TODO: stub
+	requests := req.GetRequests()
+	writeCmdOps := d.createWriteCmdOps(requests)
+	for _, op := range writeCmdOps.prewrites {
+		d.execPrewrite(aCtx, op)
+	}
+	for _, op := range writeCmdOps.commits {
+		d.execCommit(aCtx, op)
+	}
+	for _, op := range writeCmdOps.rollbacks {
+		d.execRollback(aCtx, op)
+	}
+	for _, op := range writeCmdOps.delRanges {
+		d.execDeleteRange(aCtx, op)
+	}
+	resps := make([]raft_cmdpb.Response, len(requests))
+	respPtrs := make([]*raft_cmdpb.Response, len(requests))
+	for i := 0; i < len(resps); i++ {
+		resp := &resps[i]
+		resp.CmdType = requests[i].CmdType
+		respPtrs[i] = resp
+	}
+	resp = newCmdRespForReq(req)
+	resp.Responses = respPtrs
+	if len(writeCmdOps.delRanges) > 0 {
+		result = applyResult{
+			tp:   applyResultTypeExecResult,
+			data: &execResultDeleteRange{},
+		}
+	}
+	return
+}
+
+// every commit must followed with a delete lock.
+type commitOp struct {
+	putWrite *raft_cmdpb.PutRequest
+	delLock  *raft_cmdpb.DeleteRequest
+}
+
+// a prewrite may optionally has a put Default.
+// a put default must follows a put lock.
+type prewriteOp struct {
+	putDefault *raft_cmdpb.PutRequest // optional
+	putLock    *raft_cmdpb.PutRequest
+}
+
+// a Rollback optionally has a delete Default.
+type rollbackOp struct {
+	delDefault *raft_cmdpb.DeleteRequest
+	putWrite   *raft_cmdpb.PutRequest
+	delLock    *raft_cmdpb.DeleteRequest
+}
+
+type writeCmdOps struct {
+	prewrites []prewriteOp
+	commits   []commitOp
+	rollbacks []rollbackOp
+	delRanges []*raft_cmdpb.DeleteRangeRequest
+}
+
+// createWriteCmdOps regroups requests into operations.
+func (d *applyDelegate) createWriteCmdOps(requests []*raft_cmdpb.Request) (ops writeCmdOps) {
```

Comment:
Better to add a test to cover this function.

Reply:
I'll do it in future PR.